cmd/snap/cmd_interface.go

```diff
@@ -0,0 +1,183 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2017 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package main
+
+import (
+	"fmt"
+	"io"
+	"sort"
+	"text/tabwriter"
+
+	"github.com/snapcore/snapd/client"
+	"github.com/snapcore/snapd/i18n"
+
+	"github.com/jessevdk/go-flags"
+)
+
+type cmdInterface struct {
+	ShowAttrs   bool `long:"attrs"`
+	ShowAll     bool `long:"all"`
+	Positionals struct {
+		Interface interfaceName `skip-help:"true"`
+	} `positional-args:"true"`
+}
+
+var shortInterfaceHelp = i18n.G("Lists snap interfaces")
+var longInterfaceHelp = i18n.G(`
+The interface command shows details of snap interfaces.
+
+If no interface name is provided, a list of interface names with at least
+one connection is shown, or a list of all interfaces if --all is provided.
+`)
+
+func init() {
+	addCommand("interface", shortInterfaceHelp, longInterfaceHelp, func() flags.Commander {
+		return &cmdInterface{}
+	}, map[string]string{
+		"attrs": i18n.G("Show interface attributes"),
+		"all":   i18n.G("Include unused interfaces"),
+	}, []argDesc{{
+		name: i18n.G("<interface>"),
+		desc: i18n.G("Show details of a specific interface"),
+	}})
+}
+
+func (x *cmdInterface) Execute(args []string) error {
+	if len(args) > 0 {
+		return ErrExtraArgs
+	}
+
+	if x.Positionals.Interface != "" {
+		// Show one interface in detail.
+		name := string(x.Positionals.Interface)
+		ifaces, err := Client().Interfaces(&client.InterfaceQueryOptions{
+			Names: []string{name},
+			Doc:   true,
+			Plugs: true,
+			Slots: true,
+		})
+		if err != nil {
+			return err
+		}
+		if len(ifaces) == 0 {
```

Comment:
I was looking at snap list which has similar client-side behavior. We could change the response format so that unknown interfaces can be reported but if we do so we should probably use the same approach for other listings.